### ISSUE TITLE
Allow deletion of the machines with already lowered priority.

### ISSUE THREAD

**Issue by hardikdr** [gardener/autoscaler#64]
**What would you like to be added**: There can be a case where a machine's priority is lowered but the actual machine was not terminated, probably due to the non-atomicity of the selective scale-down. 
- This can happen if the autoscaler succeeds in annotating the machine object with low priority but it somehow failed to decrease the size of the machine-deployment.

**Why is this needed**: To unblock the deletion of the machine which has already had a low priority of 1.

**Comment by hardikdr** [comment]
I think we should identify a sophisticated way to counter the non-atomicity of the selective-scaledown. Essentially there are 2 calls while scale-down:
1. Lower the annotation of the machine object.
2. Scale-down the machine-deployment by one.

If autoscaler restarts after call-1 or call-2 fail for whatever reason(even after retries), the priorities of the machines could be inconsistent. eg. one machine with a priority of 1, while others at 3. 
- This can impact the next set of scale-down decisions, based on which machine is chosen for scale-down. 

We added [this](https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/cloudprovider/mcm/mcm_manager.go#L599-L608) safety-check to ensure, we dont end-up deleting the same machine-twice.

As short-term mitigation: I think, we should remove the second check, where the autoscaler refuses to scale-down the machine which already has a priority of 1. 
- As against what we thought, with adoption, we once in a while see clusters with a machine with inconsistent priorities, leading to blocking of scale-down of one of the machines.

In long term, we should find a way to better guard autoscaler against the non-atomicity of the selective scale-down or have a better atomic-scale-down feature, which couldn't in past.

**Comment by hardikdr** [comment]
cc @prashanth26 @AxiomSamarth

**Comment by prashanth26** [comment]
Hi @hardikdr ,

Thanks for these inputs. 

> As short-term mitigation: I think, we should remove the second check, where the autoscaler refuses to scale-down the machine which already has a priority of 1.

By this you mean to remove [this part](https://github.com/gardener/autoscaler/pull/59/files#diff-fc87dde41e1a7abff0b0bc24dbabc258ac3d594a346b8f01e020b9e7f9775440R543-R546) of the code right? And this should fix the issue we see currently right?

**Comment by hardikdr** [comment]
>By this you mean to remove this part of the code right? And this should fix the issue we see currently right?

Yes, it should at least unblock the deletion of the machines with lowered priority.

We added that check, being more defensive on not allowing autoscaler to act when priorities are in an inconsistent state, but seems we are hitting it a bit more often. In general, if that check is not there, it should be self-healed, but the yet another corner case:
- If autoscaler comes up fresh, and picks up another node but the previous one, we'll have a prolonged inconsistent state, but that's anyways not worse than what we see now.

Though, I'd say it would be great to digest this issue of non-atomic selective scale-down a bit more, in general.

**Comment by prashanth26** [comment]
To resolve the problem for clusters using an older autoscaler version manually, the following would be the fix.

- Target the cluster
- `kubectl edit nodes`
- Search for taint with key as ToBeDeletedByClusterAutoscaler
- Delete such taints on the node objects.
